code/modules/mob/mob.dm

```diff
@@ -1283,10 +1283,9 @@
 /mob/proc/set_stat(new_stat)
 	if(new_stat == stat)
 		return
-	SEND_SIGNAL(src, COMSIG_MOB_STATCHANGE, new_stat)
 	. = stat
 	stat = new_stat
-
+	SEND_SIGNAL(src, COMSIG_MOB_STATCHANGE, new_stat, .)
```

Comment:
Is moving this signal necessary? Rohesie gets upset at this sort of thing.

Reply:
it seems strange to send an event signal before the change actually happens, and this structural thing is what actually caused this bug, at least generally. Fucking up a cooldown shouldn't potentially make infinite loops.

I've asked roh to comment on this, but I'm not sure why it would be structured like this

Reply:
It's to do with being able to cancel stat changes, but I never really agreed with it because then stat change signals will still fire even if they're being cancelled.

Personally, I think this is fine, I was just curious.

Reply:
Stuff like this is how we end up with COMSIG_MOB_PRE_STATCHANGE and COMSIG_MOB_STATCHANGE.

Reply:
Which could be a solution, but always felt jank.

Reply:
how does this effect being able to cancel stat changes?

Reply:
In this case it doesn't, but signals before a stat change are supposed to be wrapped with an if so the proc can early return if it is indicated to do so by signal handlers.

Reply:
>  Which could be a solution, but always felt jank.

It's the right solution, because there's no other way to solve the problem of running signals on something that's going to be cancelled.

> In this case it doesn't, but signals before a stat change are supposed to be wrapped with an if so the proc can early return if it is indicated to do so by signal handlers.

Not "supposed" to be, but rather *could* be.

Reply:
> Stuff like this is how we end up with COMSIG_MOB_PRE_STATCHANGE and COMSIG_MOB_STATCHANGE.
> Which could be a solution, but always felt jank.

I mean, sometimes you do need to do stuff after an event and sometimes you gotta do stuff before, it happens. Most implementations of it in tg code usually end up being quite janky though because they don't do it exactly before or after, usually inbetween.